File: pulsar/java/src/main/java/com/examples/hello/pulsar/producers/BatchProducerRunner.java
package com.examples.hello.pulsar.producers;

import com.examples.hello.pulsar.producers.batch.BatchProducer;
import com.examples.hello.pulsar.services.PulsarRunner;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import lombok.val;
import org.apache.pulsar.client.api.Producer;

@Slf4j
@RequiredArgsConstructor
public class BatchProducerRunner<T> extends PulsarRunner implements ProducerRunnable {
	private final Producer<T> producer;
	private final BatchProducer<T> source;

	@SneakyThrows
	@Override
	protected void callOnce() {
		val messages = source.call();
		for (val message : messages)
			producer.send(message);
		log.debug("Produced {} values (batch)", messages.size());
	}

	@Override
	public void close() throws Exception {
		producer.close();
	}
}

File: pulsar/kotlin/src/main/kotlin/com/example/hello/pulsar/producers/BatchProducerRunner.kt
package com.example.hello.pulsar.producers

import com.example.hello.pulsar.services.PulsarRunner
import org.apache.pulsar.client.api.Producer
import org.slf4j.LoggerFactory

class BatchProducerRunner<T>(
    private val producer: Producer<T>,
    private val source: () -> List<T>
) : PulsarRunner(), BatchProducerRunnable {
    override fun callOnce() {
        val messages = source.invoke()
        for (message in messages) producer.send(message)
        log.debug("Produced {} values (batch)", messages.size)
    }

    override fun close() {
        producer.close()
    }

  companion object {
    private val log = LoggerFactory.getLogger(BatchProducerRunner::class.java)
  }
}
